src/Text/Pandoc/Writers/Texinfo.hs

```diff
@@ -254,11 +254,39 @@ blockToTexinfo (Table _ blkCapt specs thead tbody tfoot) = do
                   text "@end multitable"
   return $ if isEmpty captionText
               then tableBody <> blankline
-              else text "@float" $$
+              else text "@float Table" $$
                    tableBody $$
                    inCmd "caption" captionText $$
                    text "@end float"
 
+blockToTexinfo (Figure _ (Caption _ caption) [SimpleFigure attr figCaption target]) =
+  blockToTexinfo (SimpleFigure attr (if null figCaption
+                                        then blocksToInlines caption
+                                        else figCaption)
+                                    target)
+
+blockToTexinfo (Figure _ fCaption [
+    Table attr tCaption@(Caption _ cbody) specs thead tbody tfoot]) = do
+  let caption = blocksToInlines cbody
+  captionText <- inlineListToTexinfo caption
+  blockToTexinfo (Table attr (if null captionText
+                                        then fCaption
+                                        else tCaption)
+                                    specs thead tbody tfoot)
+
+blockToTexinfo (Figure _ (Caption _ caption) body) = do
+  captionText <- inlineListToTexinfo $ blocksToInlines caption
+  content <- blockListToTexinfo body
+  return $ text ("@float" ++ floatType body) $$ content $$ (
+      if isEmpty captionText
+         then empty
+         else inCmd "caption" captionText
+    ) $$ text "@end float"
+  where
+  floatType [SimpleFigure {}] = " Figure"
+  floatType [Table {}] = " Table"
+  floatType _ = ""
+
```

Comment:
The modifications in the output here are backed by [this documentation](https://www.gnu.org/software/texinfo/manual/texinfo/html_node/_0040float.html).